Refactor taxonomic property filter logic

mariusandra opened the pull request:
## Changes

- Attempts to be a cleaner version of https://github.com/PostHog/posthog/pull/4844

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

samwinslow committed:
init TaxonomicPropertyFilter based on Tabbed variant

samwinslow committed:
InfiniteSelectResults

samwinslow committed:
infinite loader state mgmt

samwinslow committed:
set local state on select

samwinslow committed:
set by name, not id

samwinslow committed:
Merge branch 'master' into 4267-taxonomic-property-filter

samwinslow committed:
Merge branch 'master' into 4267-taxonomic-property-filter

samwinslow committed:
list and filter static values

samwinslow committed:
value search & loading anim

samwinslow committed:
allow selecting operator and value

samwinslow committed:
add cohorts support - WIP

samwinslow committed:
Fix cohort select

samwinslow committed:
default by opening to the correct tab

samwinslow committed:
Merge branch 'master' into 4267-taxonomic-property-filter

samwinslow committed:
Merge branch 'master' into 4267-taxonomic-property-filter

samwinslow committed:
extract feature flag

samwinslow committed:
add pagekey to taxonomic filter logic

samwinslow committed:
remove use of local state in taxon filter

samwinslow committed:
create keyed infiniteSelectResultsLogic

samwinslow committed:
not working: loading logic; need breakpoints and return Promise

samwinslow committed:
address minor changes

samwinslow committed:
move files to own folder

samwinslow committed:
TaxonomicPropertyFilter now controls activeTabKey

samwinslow committed:
wip: Load next chunk of results when list renders

samwinslow committed:
change pageKeys

samwinslow committed:
load initial results on mount

samwinslow committed:
set breakpoint and overscan count

samwinslow committed:
add a nextOffset state value rather than breakpt

samwinslow committed:
preserve scroll pos

samwinslow committed:
allow loading at arbitrary scroll positions

samwinslow committed:
styling

samwinslow committed:
minor cleanup

samwinslow committed:
wip: limit and offset via list accessors

samwinslow committed:
do a silly little array slice

samwinslow committed:
separate endpoint for paginated results

samwinslow committed:
use Fuse for static lists

samwinslow committed:
wip: tab titles

samwinslow committed:
counts update on search for both local and async properties

samwinslow committed:
Merge branch 'master' into 4267-taxonomic-property-filter

samwinslow committed:
revert changes to person.py

samwinslow committed:
minor improvements

mariusandra committed:
Merge branch 'master' into 4267-taxonomic-property-filter

mariusandra committed:
refactor taxonomic property filter logic WIP

mariusandra committed:
fix bugs, add search queries

mariusandra committed:
fix key error

mariusandra committed:
also show local items

mariusandra committed:
show counts

mariusandra committed:
add operator value select

mariusandra committed:
local search with fuse

mariusandra committed:
get rid of tabKey, select active tab based on filter

mariusandra committed:
various cleanup

mariusandra committed:
add simple api cache

mariusandra committed:
hide property popover -> make the popup usable

mariusandra committed:
create filter dropdown

mariusandra committed:
"add filter"

mariusandra committed:
show the right cohort

mariusandra commented:
All right, I think the refactor is done. :)

![2021-07-13 22 14 36](https://user-images.githubusercontent.com/53387/125518687-267fa0a6-a4fd-44e8-ab7b-a59ddb0cecec.gif)

There are still many things to implement for a better user experience. Some of the biggest bugs:
- up/down keys don't do anything
- the page often jumps to top for no real reason
- I had to disable the property popovers as they were just in the way all the time
- in general the UX is not yet as polished as I'd like it to be... e.g. no empty states, no bold search term highlight, etc.

That said, as this is a PR [to another PR](https://github.com/PostHog/posthog/pull/4844), it should be safe to merge in as is.

I'll continue tomorrow with the other improvements.

paolodamico commented:
Is this ready for review? Description says it's WIP and failing test

mariusandra committed:
better handling for when the filter is already in a popup

mariusandra commented:
- Forgot to update the description, but yes, ready for review. Sneaked one UX improvement in as well.
- For the test, not sure. It passed now :D.
- This PR might be hard to review because it's not comparing to master. If so, feel free to just merge it to the big PR and check it then. Otherwise, most of the interesting code lives in `frontend/src/lib/components/PropertyFilters/components/TaxonomicPropertyFilter`.
- All comments and UX feedback is absolutely welcome. Fixing those will be my focus tomorrow. The UX is far from perfect right now.

samwinslow commented on frontend/src/lib/components/PropertyFilters/components/TaxonomicPropertyFilter/infiniteListLogic.ts:
```diff
@@ -0,0 +1,204 @@
+        onRowsRendered: ({ rowInfo: { startIndex, stopIndex, overscanStopIndex } }) => {
+            if (values.isRemoteDataSource) {
+                let loadFrom: number | null = null
+                for (let i = startIndex; i < (stopIndex + overscanStopIndex) / 2; i++) {
```
Won't this load every single time another row is rendered? Fairly certain that `onRowsRendered` is potentially fired many times per second; this will result in a lot of network requests for overlapping data.

paolodamico reviewed (changes requested):
Great progress here! Reviewed mostly by testing out. Comments below.

- There's this strange extra blank space.
<img width="649" alt="" src="https://user-images.githubusercontent.com/5864173/125619058-960793e4-2919-4943-8c63-fad6c96e51c9.png">
- Initial scrolling feels off. When you scroll so your cursor is over the new component, you still can't scroll the inner component of the view unless you wait half a second (or move your cursor). The behavior is different than the event/action selector. Not sure if this is due to the extra blank space.

![image](https://user-images.githubusercontent.com/5864173/125618953-836b8469-bb0f-49a9-a42e-48826672adc9.gif)

- Z-index needs adjustment. Popover shouldn't be scrolled over the "top navbar".
- Behavior of selecting a cohort is strange, this thing is kept open after selecting one.

![image](https://user-images.githubusercontent.com/5864173/125619730-8ee1374a-358e-482e-8637-8d311a819f15.gif)

- Strange behavior when selecting one of the multiple options (the non-selected option disappears).

![image](https://user-images.githubusercontent.com/5864173/125620215-92573985-3fe8-4e92-a63f-0db7c4f399d7.gif)



### Improvements. Separate PR?
- The "All tab" seems quite important, but can't be introduced unless we add the scope icons to avoid confusion. Do we want to nail this in this PR?
- Maybe we can introduce the hotkeys (think we landed on tab) to move between tabs?

mariusandra committed:
Merge branch 'master' into taxonomic-filter-logic-refactor

mariusandra committed:
move up/down

mariusandra committed:
keyboard up/down navigation

mariusandra committed:
select on enter

mariusandra committed:
preselect selected item, fix focus jank

mariusandra committed:
autofocus on other dropdowns as well

mariusandra committed:
improve style and refactor util

mariusandra committed:
fix dumb bug

mariusandra committed:
show cohorts full screen

mariusandra committed:
fix filter alignment, type and placeholder

mariusandra committed:
filter position and responsiveness

mariusandra committed:
Merge branch 'master' into taxonomic-filter-logic-refactor

mariusandra committed:
reduce dependence on antd, fix various layout bugs

mariusandra committed:
fix responsiveness

mariusandra committed:
tooltip position

mariusandra committed:
pass on onComplete

mariusandra commented:
I hope this is ready for one more look! Quite a lot has changed code-wise, so I just closed the old PR. Makes it easier to compare to master.

1. Keyboard shortcuts now work. Including tab.
2. The filter row got some CSS-grid-powered upgrades making it more responsive. And yes, everything new works with ~360px
![2021-07-16 01 49 45](https://user-images.githubusercontent.com/53387/125871824-589aee71-0ff6-481d-b8de-f06dbdfb799d.gif)
3. The tooltips are there only when there's space. I'd be happy to work on something better for when there's no space, a "(i)" icon or something perhaps?
![2021-07-16 01 51 26](https://user-images.githubusercontent.com/53387/125871938-82bc93ee-9a16-4a35-a6ff-79060617d8d0.gif)
4. Various fixes tweaks all around.

This has gotten quite long again :)

mariusandra commented:
I noticed now that we can't autocomplete property definitions when using Clickhouse:

```
AttributeError: 'RawQuerySet' object has no attribute 'filter'
```

mariusandra committed:
hide tooltip if switching tab

mariusandra committed:
search property filters also in EE

mariusandra commented:
I fixed the property filter search, but it's not the most elegant solution. The same logic should also be applied to event definitions, eventually.

mariusandra commented:
Also, the property filter search now splits everything by a space and does an AND search, so "active feature flags" will search for `name ilike "%active%" AND name ilike '%feature%' AND name ilike '%flags%'`. I think this will give enough fuzziness that we don't really need to do the trigram search. It might speed things up though.

samwinslow commented:
Working on something using `ngramSearchCaseInsensitiveUTF8` for Clickhouse right now — it's built in 😄

mariusandra committed:
Merge branch 'master' into taxonomic-filter-logic-refactor

mariusandra committed:
add type, fix typegen write loop

mariusandra committed:
Merge branch 'fix-type-loop' into taxonomic-filter-logic-refactor

paolodamico committed:
Merge branch 'master' into taxonomic-filter-logic-refactor

paolodamico commented on posthog/api/property_definition.py:
```diff
@@ -33,10 +34,26 @@ class PropertyDefinitionViewSet(
 
     def get_queryset(self):
+        # input query
+        search_string = self.request.GET.get("search", None)
```
**Blocking**. I don't think these changes make sense because #5039 is coming in.

paolodamico commented on frontend/src/lib/components/PropertyFilters/components/TaxonomicPropertyFilter/taxonomicPropertyFilterLogic.ts:
```diff
@@ -0,0 +1,139 @@
+        tabLeft: true,
+        tabRight: true,
+        setSearchQuery: (searchQuery: string) => ({ searchQuery }),
+        setActiveTab: (activeTab: string) => ({ activeTab }),
```
Maybe we want to type `activeTab` more explicitly?

paolodamico committed:
Minor copy adjustment

You're actually searching all at once

paolodamico reviewed (commented):
We still have a few visual shortcomings but as this PR is becoming too large and code is feature flagged I would maybe consider getting this in and fixing in a separate PR. Though I would say this is not ready for a release to real users.

1. IMO the popover animation is actually detrimental. Compare to the event/action selector that opens instantly, feels faster.
2. You already noted this but we have no empty state. Probably we need a simple design here? CC @clarkus 
3. If I press enter on a property, the whole thing gets submitted and I don't get a chance to select a value.
4. The search string is not cleared after closing the popover (either cancelling or selecting a property). Quite strange particularly when you open a second "Add filter" popover that has nothing to do with the first one.
5. Popover content gets cut off (exceeds container).
    <img width="1355" alt="" src="https://user-images.githubusercontent.com/5864173/126046413-e08a5768-f5cb-42a4-a76d-7e4f86db7663.png">
6. On smaller screen widths (~600px), the popover exceeds the window size and you can actually scroll out of bounds.
    <img width="786" alt="" src="https://user-images.githubusercontent.com/5864173/126046428-e3c143fe-4e2b-4525-9adf-f4f92047dc6f.png">
7. After selecting a property, when you should type/select values, the layout gets cut off and unnecessarily multiline (tested full screen on 13" MB)
<img width="574" alt="Screenshot 2021-07-17 at 1 12 58 PM" src="https://user-images.githubusercontent.com/5864173/126046461-969236bf-b2ef-4f20-b58c-7c2c4d98a8d6.png">
8. Selecting an option makes **the other** options disappear.
![image](https://user-images.githubusercontent.com/5864173/126046482-ae6a3b73-0ab9-49ce-9a0c-544ff322010a.gif)

mariusandra committed:
fix outside click handler

mariusandra committed:
popups with popper

mariusandra committed:
add ant-select as exception for popup outside clicks

mariusandra committed:
simplify deps

mariusandra committed:
always highlight the first item

mariusandra committed:
use new popup with taxonomic filter

mariusandra committed:
fix PropertyFilterButton ref

mariusandra committed:
clear search on select

mariusandra committed:
placeholder for value input box

mariusandra committed:
add empty state

mariusandra commented:
Thank you for the review @paolodamico ! I've addressed some of the issues and more to come. But first, when I marked this as ready to review, the popups definitely didn't look as broken as they now were. Not sure what happened there, perhaps a result of a merge? 

Comments on the review points:
1. I kept all the animations as they were, but indeed, it's not a nice animation. I created a new popup component using [react-popper](https://github.com/popperjs/react-popper): "*Popper is not a tooltip library, it's a positioning engine to be used to build features such as (but not restricted to) tooltips.*". I used it for both the previously slow "global filters" popups and just the general select dropdown.
2. Added the default antd empty state.
3. TODO
4. Done
5. Fixed
6. Fixed
7. Somewhat fixed. No longer cut off, but still multiline.
8. I believe this is really annoying and I want to change it, though I would prefer not to touch the `<PropertyFilterValue />` component now. 
9. (backend API search) - yes, will review/merge the trigram stuff separately
10. (`activeTab` typing) - yes, TODO

mariusandra committed:
only open filter downwards

mariusandra committed:
move listTooltip postition into useState

mariusandra committed:
list tooltip position after opening and resize events

paolodamico reviewed (approved):
Looks like the most critical issues have been addressed, let's merge this and continue improving it in a separate PR

mariusandra commented:
The separate PR is already here: https://github.com/PostHog/posthog/pull/5189
I've been making solid progress and should have something that doesn't suck in a few hours :)

mariusandra committed:
Merge branch 'master' into taxonomic-filter-logic-refactor
